### PR TITLE
Place both contract code and state behind link

### LINKED ISSUE #333: Port VM1 contract code behind link to VM2 and add state behind link in VM2 contract
**Describe what you want implemented** 
Contract code was put behind link in VM1, this needs to be ported to VM2. Also, it is beneficial to put contract state behind link as well.

**Describe "Why" this is needed**
When state changes, contract code does not need to be rewritten when re-serialising the state. State behind link allows contracts to share same state, e.g., in a hamt node.

**Describe alternatives you've considered**
We could keep contract code and state without links, but this is less efficient as every change in state causes code to be rewritten.

**Additional context**
Not having a link also causes a storage size bloat. That's why it was introduced in VM1.

### PR DESCRIPTION
**Pull request by ureeves**
Resolves #333

### CODE REVIEW

**Review by krl — APPROVED** [review]
Looks good, and good that you removed the trait as well.

**Review by miloszm — CHANGES_REQUESTED** [review]
Looks really good, glad ContractRef is gone. Have some remarks and questions, but overall LGTM. Biggest question - why transact does not have similar changes?

File: src/call_context.rs [hunk 1]
```diff
@@ -144,10 +143,15 @@ impl<'a> CallContext<'a> {
...
-            )?;
+            let bytecode = match contract {
+                MaybeArchived::Memory(m) => m.bytecode(),
+                MaybeArchived::Archived(a) => {
+                    self.store.get(a.bytecode()).as_ref()
+                }
+            };
+
+            let module =
+                compile_module(bytecode, self.state.get_module_config())?;
```
[the remark this answers is not in the document]
Comment (ureeves) [reply src/call_context.rs:155]: I can't implement anything for `MaybeUnarchived` - unless we use a trait (ala `ContractRef`). The best I can do is encapsulate the logic for getting stuff from `ArchivedContract`.

File: test_runner/src/main.rs [hunk 2]
```diff
@@ -317,7 +309,7 @@ fn confirm_stack(source_path: impl AsRef<str>) -> Result<(), Box<dyn Error>> {
 
     let mut gas = GasMeter::with_limit(100_000_000_000);
     for i in 0..STACK_TEST_SIZE {
-        let ii = network
+        let ii = *network
```
[the remark this answers is not in the document]
Comment (ureeves) [reply test_runner/src/main.rs:312]: I can make a separate PR for this, but `make test-session` was not compiling an I decided to fix it.

[the remark this answers is not in the document]
Comment (miloszm) [reply test_runner/src/main.rs:312]: I see, that's ok, test-session is held up on issue 350, hence it is not maintained atm

**Review by miloszm — APPROVED** [review]
thanks for your responses, looks really good

**Pull request merged**